### PR TITLE
[SPARK-34786][SQL] Read Parquet unsigned int64 logical type that stored as signed int64 physical type to decimal(20, 0)

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

A companion PR for SPARK-34817, when we handle the unsigned int(<=32) logical types. In this PR, we map the unsigned int64 to decimal(20, 0) for better compatibility. 

### Why are the changes needed?

Spark won't have unsigned types, but spark should be able to read existing parquet files written by other systems that support unsigned types for better compatibility. 

### Does this PR introduce _any_ user-facing change?

yes, we can read parquet uint64 now

### How was this patch tested?

new unit tests

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu @HyukjinKwon please take a look at this PR thanks

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetDictionary.java [hunk 1]
```diff
@@ -61,6 +63,14 @@ public double decodeToDouble(int id) {
 
   @Override
   public byte[] decodeToBinary(int id) {
-    return dictionary.decodeToBinary(id).getBytes();
+    if (needTransform) {
+      // For unsigned int64, it stores as dictionary encoded signed int64 in Parquet
+      // whenever dictionary is available.
+      // Here we lazily decode it to the original signed long value then convert to decimal(20, 0).
+      long signed = dictionary.decodeToLong(id);
+      return new BigInteger(Long.toUnsignedString(signed)).toByteArray();
```
Comment (cloud-fan) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetDictionary.java:71]: Does the binary of `BigInteger` can be read back as `BigDecimal`?

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java [hunk 2]
```diff
@@ -613,6 +631,8 @@ private void readLongBatch(int rowId, int num, WritableColumnVector column) thro
       defColumn.readLongs(
         num, column, rowId, maxDefLevel, (VectorizedValuesReader) dataColumn,
         DecimalType.is32BitDecimalType(column.dataType()));
+    } else if (originalType == OriginalType.UINT_64) {
+      defColumn.readUnsignedLongs(num, column, rowId, maxDefLevel, (VectorizedValuesReader) dataColumn);
```
Comment (cloud-fan) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java:635]: can we add some comments here as well?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala [hunk 3]
```diff
@@ -98,9 +98,6 @@ class ParquetToSparkSchemaConverter(
     def typeString =
       if (originalType == null) s"$typeName" else s"$typeName ($originalType)"
 
-    def typeNotSupported() =
-      throw QueryCompilationErrors.parquetTypeUnsupportedError(typeString)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala:102]: can we remove QueryCompilationErrors.parquetTypeUnsupportedError?

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetDictionary.java (hunk 1 above)
Comment (yaooqinn) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetDictionary.java:71]: use `BigInteger` for caller side
 https://github.com/apache/spark/blob/a418548dad57775fbb10b4ea690610bad1a8bfb0/sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java#L371

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java (hunk 2 above)
Comment (yaooqinn) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java:635]: OK~

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala:102]: good catch!

**Review by cloud-fan — APPROVED** [review]

**Comment by yaooqinn** [comment]
Thanks, merged to master

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala [hunk 4]
```diff
@@ -144,7 +141,7 @@ class ParquetToSparkSchemaConverter(
         originalType match {
           case INT_64 | null => LongType
           case DECIMAL => makeDecimalType(Decimal.MAX_LONG_DIGITS)
-          case UINT_64 => typeNotSupported()
+          case UINT_64 => DecimalType.LongDecimal
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala:144]: I don't think it's related to `DecimalType.LongDecimal`. We want it to store uint64, so the precision we need is `java.lang.Long.toUnsignedString(-1).length`, which is 20.

Let's make it more explicit
```
// The precision to hold the largest unsigned long is: `java.lang.Long.toUnsignedString(-1).length` = 20
case UINT_64 => DecimalType(20, 0)
```

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala:144]: OK, I will send a followup

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala [hunk 5]
```diff
@@ -459,6 +461,40 @@ class ParquetIOSuite extends QueryTest with ParquetTest with SharedSparkSession
...
+            .append("a", i % 100L)
+          writer.write(group)
+        }
+        writer.close()
+      }
+
+      withTempDir { dir =>
+        val path = new Path(dir.toURI.toString, "part-r-0.parquet")
+        makeRawParquetFile(path)
+        readParquetFile(path.toString) { df =>
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetIOSuite.scala:488]: can we also test with user-specified schema? What if the user-specified schema is not `decimal(20, 0)`? Do we return wrong result or fail?

**Pull request closed without merging**